File: leetcode/Javascript/001_two_sum.js
/*
Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    var ans = [0, 0];
    var map = {};
    for (var i = 0; i < nums.length; i++) {
        var tmp = target - nums[i];
        if (typeof map[tmp] != "undefined") {
            ans[0] = map[tmp];
            ans[1] = i;
            return ans;
        } else {
            map[nums[i]] = i;
        }
    }
    return ans;
};

File: leetcode/cpp/001_two_sum.cpp
/*

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].

*/

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> dic;
        vector<int> ans;
        for (int i = 0; i < nums.size(); i++) {
            if ( dic.find(target - nums[i]) != dic.end() ) {
                ans.push_back(i);
                ans.push_back(dic[target - nums[i]]);
                return ans;
            }
            dic[nums[i]] = i;
        }
        ans.push_back(-1);
        ans.push_back(-1);
        return ans;
    }
};
